### ISSUE TITLE
Brucker BCF reader fails to load some files

### ISSUE THREAD

**Issue by francisco-dlp** [hyperspy/hyperspy#1502]
This is the error:

```python

In [1]: import hyperspy.api as hs

In [2]: hs.load("map2-data.bcf")
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-2-2cb6879a9d88> in <module>()
----> 1 hs.load("map2-data.bcf")

/home/fjd29/Python/hyperspy3/hyperspy/io.py in load(filenames, signal_type, stack, stack_axis, new_axis_name, lazy, **kwds)
    229             objects = [load_single_file(filename, lazy=lazy,
    230                                         **kwds)
--> 231                        for filename in filenames]
    232 
    233         if preferences.Plot.plot_on_load:

/home/fjd29/Python/hyperspy3/hyperspy/io.py in <listcomp>(.0)
    229             objects = [load_single_file(filename, lazy=lazy,
    230                                         **kwds)
--> 231                        for filename in filenames]
    232 
    233         if preferences.Plot.plot_on_load:

/home/fjd29/Python/hyperspy3/hyperspy/io.py in load_single_file(filename, signal_type, **kwds)
    275                                 reader=reader,
    276                                 signal_type=signal_type,
--> 277                                 **kwds)
    278 
    279 

/home/fjd29/Python/hyperspy3/hyperspy/io.py in load_with_reader(filename, reader, signal_type, **kwds)
    284     lazy = kwds.get('lazy', False)
    285     file_data_list = reader.file_reader(filename,
--> 286                                         **kwds)
    287     objects = []
    288 

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in file_reader(filename, select_type, index, downsample, cutoff_at_kV, instrument, lazy)
   1105 
   1106     # objectified bcf file:
-> 1107     obj_bcf = BCF_reader(filename)
   1108     if select_type == 'image':
   1109         return bcf_imagery(obj_bcf)

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in __init__(self, filename)
    801         header_file = self.get_file('EDSDatabase/HeaderData')
    802         header_byte_str = header_file.get_as_BytesIO_string().getvalue()
--> 803         self.header = HyperHeader(header_byte_str)
    804         self.hypermap = {}
    805 

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in __init__(self, xml_str)
    587         #self.channel_factors = {}
    588         self.spectra_data = {}
--> 589         self._set_sum_edx(root)
    590 
    591     def _set_sem(self, root):

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in _set_sum_edx(self, root)
    683             #                                         str(i))[0])
    684             self.spectra_data[i] = EDXSpectrum(root.xpath("SpectrumData" +
--> 685                                                           str(i))[0].ClassInstance)
    686 
    687     def estimate_map_channels(self, index=0):

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in __init__(self, spectrum)
    490         self.hardware_metadata = json.loads(json.dumps(hardware_header,
    491                                                        cls=ObjectifyJSONEncoder))
--> 492         self.amplification = self.hardware_metadata['Amplification']  # USED
    493 
    494         # map stuff from detector xml branch

KeyError: 'Amplification'

```

Commenting out that line leads to more missing metadata errors. I can provide a file if needed to debug.

**Comment by francisco-dlp** [comment]
Using HyperSpy 1.0 I get a different error:


```python

In [1]: import hyperspy.api as hs
unbcf_fast library is not present...
Falling back to slow python only backend.

In [2]: hs.load("map2-data.bcf")
WARNING:hyperspy.io_plugins.bcf:spectrum have no dead time records...
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-2cb6879a9d88> in <module>()
----> 1 hs.load("map2-data.bcf")

/home/fjd29/Python/hyperspy3/hyperspy/io.py in load(filenames, signal_type, stack, stack_axis, new_axis_name, mmap, mmap_dir, **kwds)
    177             objects = [load_single_file(filename,
    178                                         **kwds)
--> 179                        for filename in filenames]
    180 
    181         if hyperspy.defaults_parser.preferences.Plot.plot_on_load:

/home/fjd29/Python/hyperspy3/hyperspy/io.py in <listcomp>(.0)
    177             objects = [load_single_file(filename,
    178                                         **kwds)
--> 179                        for filename in filenames]
    180 
    181         if hyperspy.defaults_parser.preferences.Plot.plot_on_load:

/home/fjd29/Python/hyperspy3/hyperspy/io.py in load_single_file(filename, signal_type, **kwds)
    221                                 reader=reader,
    222                                 signal_type=signal_type,
--> 223                                 **kwds)
    224 
    225 

/home/fjd29/Python/hyperspy3/hyperspy/io.py in load_with_reader(filename, reader, signal_type, **kwds)
    229                      **kwds):
    230     file_data_list = reader.file_reader(filename,
--> 231                                         **kwds)
    232     objects = []
    233 

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in file_reader(filename, select_type, index, downsample, cutoff_at_kV)
    977     """
    978     #objectified bcf file:
--> 979     obj_bcf = BCF_reader(filename)
    980     if select_type == 'image':
    981         return bcf_imagery(obj_bcf)

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in __init__(self, filename)
    699         header_file = self.get_file('EDSDatabase/HeaderData')
    700         header_byte_str = header_file.get_as_BytesIO_string().getvalue()
--> 701         self.header = HyperHeader(header_byte_str)
    702         self.hypermap = {}
    703 

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in __init__(self, xml_str)
    518         self.channel_factors = {}
    519         self.spectra_data = {}
--> 520         self._set_sum_edx(root)
    521 
    522     def _set_sem(self, root):

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in _set_sum_edx(self, root)
    594                                                                     str(i))[0])
    595             self.spectra_data[i] = EDXSpectrum(root.xpath("SpectrumData" +
--> 596                                                        str(i))[0].ClassInstance)
    597 
    598     def estimate_map_channels(self, index=0):

/home/fjd29/Python/hyperspy3/hyperspy/io_plugins/bcf.py in __init__(self, spectrum)
    439             _logger.warning('spectrum have no dead time records...')
    440         self.zeroPeakPosition = int(
--> 441                       spectrum.TRTHeaderedClass.ClassInstance.ZeroPeakPosition)
    442         self.amplification = int(
    443                       spectrum.TRTHeaderedClass.ClassInstance.Amplification)

src/lxml/lxml.objectify.pyx in lxml.objectify.ObjectifiedElement.__getattr__ (src/lxml/lxml.objectify.c:4279)()

src/lxml/lxml.objectify.pyx in lxml.objectify._lookupChildOrRaise (src/lxml/lxml.objectify.c:7059)()

AttributeError: no such child: ZeroPeakPosition
```

**Comment by sem-geologist** [comment]
hmm... I need to refresh my memory..
1. do hyperspy use amplification, why did I hard coded lookup for it?
2. amplification is one of base Bruker SDD pulse processor attributes, why it is missing?

for 1 I will review the code and to understand why it is hard coded,
maybe it is there from my private implementation of poor-man FWHM estimation from zero peak, which were left out from implementing in hyperspy.
for 2, @francisco-dlp could you tell me more about the instrument (especially how old detector is, model, or even better would be the file  sample) 

Yes please I want to see the file to debug this issue.
Or if you dont want to send whole file, You could extract xml part at least and send that. (I allready showed how to do it in hyperspy/rosettasciio#239, but below I show again how to do that:
```python
from hyperspy.io_plugins import bcf
from lxml import etree

sfs = bcf.SFS_reader('your_bcffile_path.bcf')  # here we get the single file system handler
# all file items in this singlefilesystem class instance is held inside
# dictionary hierarchy, we fetch the header:
header = sfs.vfs['EDSDatabase']['HeaderData']  
#the items in the tree have special method which allows to get selection as BytesIO object:
bstring = header.get_as_BytesIO_string()
#rewind:
bstring.seek(0)
#the final steps to print the xml to file for examination in text editor:
xml = etree.fromstring(bstring.read())
xml_root = xml.getroottree()
xml_root.write('header_output.xml', pretty_print=True)
# without pretty_print everything would be in the single line
```

**Comment by francisco-dlp** [comment]
@sem-geologist, you can [download the file from here](https://dl.dropboxusercontent.com/u/479148/map2-data.bcf).

**Comment by sem-geologist** [comment]
Oh, You have nice piece of newer technology which were not tested properly (Only on the samples provided with that demo pascal code were checked) :)

**Comment by francisco-dlp** [comment]
:)

**Comment by sem-geologist** [comment]
It is probably good time to say what are the differences between the version 1 and version 2 bcf, there are many goodies in the version 2 file which would useful:

- bcf v1 xml part had the node for pixel time, but the many files I had inspected were always filled with 0 arrays of image size. BCF v2 brings in the change – those arrays now are filled with non zero data (the question is: is it real time?). It probably brings no advantage to TEM users as the current there is not such high, but for SEM users it could mean very big difference.
- bcf2 additionally contains overview image, it was not present in the bcf v1 (what is it? It is the picture on the left in the Esprit which size can differ from mapping size if only mapping the selected smaller rectangle than whole picture). If to implement it, how it should be output? as additional signal? Should it contain marker of the hypercube place?
- bcf2 contains additional table with pixel offsets for hypercube – this would allow to have random access (don't know if that would be useful in hyperspy, it needs the case to think up the implementation). If case would be thought up, we could implement generation of such tables for version 1 as well.

**Comment by francisco-dlp** [comment]
+1 for outputting the overview image as an extra signal with a permanent marker.

**Comment by sem-geologist** [comment]
ok, but how to do that? is somewhere those permanent markers documented, I mean how to insert them with dict which is passed to general hyperspy loader?

**Comment by francisco-dlp** [comment]
Just set the ``permanent`` keyword argument to ``True`` when creating the marker. Documented in the [User Guide](http://hyperspy.org/hyperspy-doc/current/user_guide/visualisation.html#plot-markers).

**Comment by sem-geologist** [comment]
Ok, I have got passed the problem with `Amplification` which were searched in wrong place (the nodes were statically unpacked, in repaired implementation it will use xpath).

But I am running into next problem: date, time.

What is the problem?:
bcf Xml have date time in many places:
under main header, under the image, under the header for some settings .... 
Date time are not the same there (some marks very beginning of acquiring of hypermap other the end of acquisition).
In previous implementation I used time from sum spectra, as that spectra is updated whole time during the acquisition, and its date/time always points to end of acquisition. It worked also with version 2 sample, but in this new file Date time is not present there.

It was actually never discussed what time /date format should io_plugin provide to hyperspy metadata? beginning or finishing time and date?

**Comment by sem-geologist** [comment]
Oh, actually I sorted out:
All the times present in the xml (with exception maybe with line selection) represents time of data acquisition. The main header represent when hypermap were started to be acquired, and it is same as the image acquisition, if the time is present in the spectrum there it is one sec earlier (I guess as the spectrum object gets instantiated before the construction of whole bcf object). So to be closest as possible I should take it not from spectra, but from image. Overview image have its own time stamp.

@francisco-dlp I hear that it is possible to select few ROIs for mapping in Esprit and save whole set into one bcf. Can you check that? 

I already can pass through the image loading, but get segfault for hypercube :/

**Comment by francisco-dlp** [comment]
@sem-geologist, I didn't acquire the dataset and I don't have access to the machine. I will ask the person who gave me the dataset to try to get another one with several ROIs for you if that's possible.

**Comment by magnunor** [comment]
@sem-geologist, I am implementing loading of permanent markers for DM-files (https://github.com/hyperspy/hyperspy/pull/1491). You should be able to adapt the code there for making markers for the BCF files as well.

**Comment by sem-geologist** [comment]
I think I will add the marker and overview picture in the other PR, as those are not bug fixes but feature addition. Anyway @magnunor, thank You.

**Comment by francisco-dlp** [comment]
@sem-geologist has submitted a fix hyperspy/hyperspy#1512
